src/unix/linux.c

```diff
@@ -298,6 +545,140 @@ int uv__io_check_fd(uv_loop_t* loop, int fd) {
+  if (head == *iou->sqtail + 1)
+    return NULL;  /* No room in ring buffer. */
+
+  slot = *iou->sqtail & iou->sqmask;
+  iou->sqarray[slot] = slot;  /* Identity mapping of index -> sqe. */
+
+  sqe = iou->sqe;
+  sqe = &sqe[slot];
+  memset(sqe, 0, sizeof(*sqe));
+  sqe->user_data = (uintptr_t) req;
+
+  /* Pacify uv_cancel(). */
+  req->work_req.loop = loop;
+  req->work_req.work = NULL;
+  req->work_req.done = NULL;
+  QUEUE_INIT(&req->work_req.wq);
+
+  uv__req_register(loop, req);
+  iou->in_flight++;
+  iou->to_submit++;
+
+  return sqe;
+}
+
+
+int uv__iou_fs_fsync_or_fdatasync(uv_loop_t* loop,
+                                  uv_fs_t* req,
+                                  uint32_t fsync_flags) {
+  uv__loop_internal_fields_t* lfields;
+  struct uv__io_uring_sqe* sqe;
+  struct uv__iou* iou;
+
+  lfields = uv__get_internal_fields(loop);
+  iou = &lfields->iou;
+
+  sqe = uv__iou_get_sqe(iou, loop, req);
+  if (sqe == NULL)
+    return 0;
+
+  /* Little known fact: setting seq->off and seq->len turns
+   * it into an asynchronous sync_file_range() operation.
+   */
+  sqe->fd = req->file;
+  sqe->fsync_flags = fsync_flags;
+  sqe->opcode = UV__IORING_OP_FSYNC;
+
+  atomic_store_explicit((_Atomic uint32_t*) iou->sqtail,
+                        *iou->sqtail + 1,
+                        memory_order_release);
+
+  return 1;
+}
+
+
+int uv__iou_fs_read_or_write(uv_loop_t* loop,
+                             uv_fs_t* req,
+                             int is_read) {
+  uv__loop_internal_fields_t* lfields;
+  struct uv__io_uring_sqe* sqe;
+  struct uv__iou* iou;
+
+  lfields = uv__get_internal_fields(loop);
+  iou = &lfields->iou;
+
+  sqe = uv__iou_get_sqe(iou, loop, req);
+  if (sqe == NULL)
+    return 0;
+
+  sqe->addr = (uintptr_t) req->bufs;
+  sqe->fd = req->file;
+  sqe->len = req->nbufs;
+  sqe->off = req->off < 0 ? -1 : req->off;
+  sqe->opcode = is_read ? UV__IORING_OP_READV : UV__IORING_OP_WRITEV;
+
+  atomic_store_explicit((_Atomic uint32_t*) iou->sqtail,
+                        *iou->sqtail + 1,
+                        memory_order_release);
+
+  return 1;
+}
+
+
+static void uv__poll_io_uring(uv_loop_t* loop,
+                              struct uv__iou* iou,
+                              int ringfd,
+                              int events) {
+  struct uv__io_uring_cqe* cqe;
+  struct uv__io_uring_cqe* e;
+  uv_fs_t* req;
+  uint32_t head;
+  uint32_t tail;
+  uint32_t mask;
+  uint32_t i;
+
+  head = *iou->cqhead;
+  tail = atomic_load_explicit((_Atomic uint32_t*) iou->cqtail,
+                              memory_order_acquire);
+  mask = iou->cqmask;
+  cqe = iou->cqe;
+
+  for (i = head; i != tail; i++) {
+    e = &cqe[i & mask];
+
+    req = (uv_fs_t*) (uintptr_t) e->user_data;
+    assert(req->type == UV_FS);
+
+    uv__req_unregister(loop, req);
+    iou->in_flight--;
+
+    /* io_uring stores error codes as negative numbers, same as libuv. */
+    req->result = e->res;
+    req->cb(req);
```

Comment:
For metrics consistency will need to add `uv__metrics_update_idle_time(loop);` before this.

It would also be helpful if you could track the number of reqs that are processed and update `uv__metrics_inc_events()` as well. I don't immediately see how we would track `uv__metrics_inc_events_waiting()`.

I have this branch running locally. Will write tests to verify metrics with this specific case.

Reply:
Actually, now I'm not sure about the call to `uv__metrics_update_idle_time()` because isn't it possible that the call to `epoll_pwait()` still idles? I'll just write the tests to verify.

Reply:
> isn't it possible that the call to epoll_pwait() still idles?

Yes, until completion events become available in the ring buffer.

Reply:
I think I can at least fix `metrics.loop_count` with the patch below. The remaining thorny issue is `metrics.events`.
<details>

```diff
diff --git a/src/unix/linux.c b/src/unix/linux.c
index dfe75c50..654bd1a0 100644
--- a/src/unix/linux.c
+++ b/src/unix/linux.c
@@ -700,6 +700,7 @@ void uv__io_poll(uv_loop_t* loop, int timeout) {
   sigset_t* sigmask;
   sigset_t sigset;
   uint64_t base;
+  int have_iou_events;
   int have_signals;
   int nevents;
   int count;
@@ -850,6 +851,7 @@ void uv__io_poll(uv_loop_t* loop, int timeout) {
       goto update_timeout;
     }

+    have_iou_events = 0;
     have_signals = 0;
     nevents = 0;

@@ -876,6 +878,7 @@ void uv__io_poll(uv_loop_t* loop, int timeout) {

       if (fd == iou->ringfd) {
         uv__poll_io_uring(loop, iou, fd, pe->events);
+        have_iou_events = 1;
         continue;
       }

@@ -950,6 +953,9 @@ void uv__io_poll(uv_loop_t* loop, int timeout) {
     loop->watchers[loop->nwatchers] = NULL;
     loop->watchers[loop->nwatchers + 1] = NULL;

+    if (have_iou_events != 0)
+      return;  /* Event loop should cycle now so don't poll again. */
+
     if (have_signals != 0)
       return;  /* Event loop should cycle now so don't poll again. */

```

</details>

Reply:
`loop_count` is taken care of somewhere else. That shouldn't be affected by any of the changes you made here.

Regarding `events`, I didn't understand the patch correctly. It should work if you changed `uv__poll_io_uring()` to return the number of times `cb()` is called, and change its call location to:
```c
nevents += uv__poll_io_uring(loop, iou, fd, pe->events);
```
and everything else will be taken care of since `nevents` is used to calculate both `events` and `events_waiting`.

Your change with `have_iou_events` will fix the `idle_time` calculation, but you'll still need to add `uv__metrics_update_idle_time(loop);` above the call here to `req->cb(req);`.

side note: I'm looking forward to node getting v1.45.0 in so I can show how useful these metrics are :)

Reply:
I have it working now! I'll mark the discussion as resolved.

Reply:
Thanks so much for taking care of that.